R/zchunk_LA100.IEA_downscale_ctry.R

```diff
@@ -0,0 +1,308 @@
+                                                             "Former Yugoslavia (if no detail)",
+                                                             "Other Africa",
+                                                             "Other Non-OECD Americas",
+                                                             "Other Asia"))
+      IEA_ctry_single <- filter(IEA_ctry, ! IEA_ctry %in% IEA_ctry_composite$IEA_ctry)
+
+      # Split IEA energy balances into table of single countries and composite regions
+      # (keeping only desired composite regions)
+      L100.IEAcomposite <- filter(L100.IEAfull, COUNTRY %in% IEA_ctry_composite$IEA_ctry)
+      L100.IEAsingle <- filter(L100.IEAfull, COUNTRY %in% IEA_ctry_single$IEA_ctry)
+      L100.IEAfull %>%
+        filter(COUNTRY %in% IEA_ctry_single$IEA_ctry) %>%
+        left_join_error_no_match(select(IEA_ctry_single, IEA_ctry, iso),
+                                 by = c("COUNTRY" = "IEA_ctry")) ->
+        L100.IEAsingle
+
+      # Subset countries that are being downscaled in certain years using historical
+      # energy data in a specified year. Former Soviet Union and Yugoslavia: use specified
+      # flows for each product. The IEA dataset has many inter-sectoral inconsistencies between
+      # the USSR and separated countries thereafter, resulting in unrealistic fuel shares.
+      USSR_YUG_SPLIT_YEAR <- 1990
+      USSR_YUG_YEARS <- HISTORICAL_YEARS[HISTORICAL_YEARS < USSR_YUG_SPLIT_YEAR]
+      POST_USSR_YUG_YEARS_IEA <- HISTORICAL_YEARS[HISTORICAL_YEARS >= USSR_YUG_SPLIT_YEAR]
+      L100.USSR_Yug <- filter(L100.IEAcomposite, COUNTRY %in% c("Former Soviet Union (if no detail)",
+                                                                "Former Yugoslavia (if no detail)"))
+
+      # Re-map the "if no detail" forms of coal in the historical years prior to the relevant
+      # coal types for matching with the more recent years (89-107)
+      NO_DETAIL_COAL_YEARS <- as.character(intersect(1971:1977, HISTORICAL_YEARS))
+      if(length(NO_DETAIL_COAL_YEARS)) {
+        # Hard coal needs to be split proportionally between coking coal and other bituminous coal
+        # in order to minimize bias from different country-wise shares of the two fuel types.
+        # Note that anthracite is not considered in these regions.
+        # NOTE: using round() to avoid NA's for any values whose base value is >1e6. This seems to work
+        prod <- L100.USSR_Yug$PRODUCT
+        prod_obc <- prod == "Other bituminous coal"
+        prod_hcind <- prod == "Hard coal (if no detail)"
+        prod_cc <- prod == "Coking coal"
+        L100.USSR_Yug[prod_obc, NO_DETAIL_COAL_YEARS] <-
+          L100.USSR_Yug[prod_hcind, NO_DETAIL_COAL_YEARS ] *
+          round(L100.USSR_Yug$`1978`[prod_obc] /
+                  (L100.USSR_Yug$`1978`[prod_cc] + L100.USSR_Yug$`1978`[prod_obc] + 1e-3), digits = 2)
```

Comment:
Maybe we could filter and set prod_cc and prod_obc to 0 when prod_hcind is 0, and do the share without 1e-3 so as not to affect the valid share calculation.